File: RetractableSidebarWindow/sidebar.cpp
#include "sidebar.h"
#include <QLibraryInfo>
Sidebar::Sidebar(QWidget *parent, int initialWidth)
    : QWidget{parent},
      animation{new QPropertyAnimation(this, "size")},
      verticalLayout{new QVBoxLayout(this)}
{
    this->verticalLayout->setSpacing(4);
    this->verticalLayout->setContentsMargins(5, 5, 5, 5);
    this->animation->setDuration(150);
    this->borderLinePen.setWidth(2);

    connect(this->animation, &QPropertyAnimation::finished, this, [&]
            {
        if(this->initialSize.height()!=this->window()->height())
        {
            //动画结束,高度和父窗口不等,说明动画进行的过程中改变了窗口大小,需重新设置高度
            this->initialSize.rheight()=this->window()->height();
            this->endSize.rheight() = this->initialSize.rheight();
            this->animation->setStartValue(this->initialSize);
            this->animation->setEndValue(this->endSize);
            this->resize(this->width(),this->window()->height());
        }
        //设置展开状态和动画运行状态,根据展开状态发射相应信号
        this->isExpanded = !this->isExpanded;
        this->isAnimationRunning = false;
        if(this->isExpanded)
            emit this->expandFinished();
        else
            emit this->shrinkFinished();
        emit this->finished(); });

    // 动画起始高度和父窗口高度一样,宽度为初始宽度,之后不再更改宽度
    this->initialSize.rwidth() = initialWidth;
    this->initialSize.rheight() = this->window()->height();

    // 动画结束高度和动画起始高度始终相等,动画结束宽度为动画起始宽度加上展开增加宽度
    this->endSize.rwidth() = this->initialSize.rwidth() + 270;
    this->endSize.rheight() = this->initialSize.rheight();

    this->resize(initialWidth, this->window()->height());
}

Sidebar::~Sidebar()
{
    this->animation->deleteLater();
}

void Sidebar::autoExpand()
{
    this->expand();
    this->shrink();
}

void Sidebar::expand()
{
    // 当前不是展开状态
    if (!this->isExpanded && !this->isAnimationRunning)
    {
        this->animation->setDirection(QAbstractAnimation::Direction::Forward); // 正向
        this->animation->start();
        this->isAnimationRunning = true;
        emit this->expandStart();
    }
}

void Sidebar::shrink()
{
    if (this->isExpanded && !this->isAnimationRunning)
    {
        this->animation->setDirection(QAbstractAnimation::Direction::Backward); // 逆向
        this->animation->start();
        this->isAnimationRunning = true;
        emit this->shrinkStart();
    }
}

void Sidebar::addItem(QLayoutItem *item)
{
    this->verticalLayout->addItem(item);
}

void Sidebar::setIncreasedWidth(quint32 increasedWidth)
{
    // increasedWidth - (this->endSize.rwidth() - this->initialSize.rwidth());
    this->endSize.rwidth() += increasedWidth - this->endSize.rwidth() + this->initialSize.rwidth();
}

void Sidebar::setExpandTime(int ms)
{
    this->animation->setDuration(ms);
}

void Sidebar::setBackgroundBrush(const QBrush &brush)
{
    this->backgroundBrush = brush;
}

void Sidebar::setBorderLinePen(const QPen &pen)
{
    this->borderLinePen = pen;
}

int Sidebar::childrenCumulativeHeight()
{
    int cumulativeHeight = 0;                                                                  // 累计高度

#if (QT_VERSION >= QT_VERSION_CHECK(6,3,0))
    auto children(this->findChildren<QWidget *>(Qt::FindChildOption::FindDirectChildrenOnly)); // 查找直接子控件
#else
    auto children(this->findChildren<QWidget *>(QRegularExpression(QString(R"([\s\S]+)")),Qt::FindChildOption::FindDirectChildrenOnly));
#endif
    for (auto &child : children)
        cumulativeHeight += child->height(); // 累加直接子控件高度

    // 累加布局边距和控件间距
    cumulativeHeight += layout()->contentsMargins().top();
    cumulativeHeight += layout()->contentsMargins().bottom();
    cumulativeHeight += layout()->spacing() * (children.count() - 1);

    return cumulativeHeight;
}

void Sidebar::addWidget(QWidget *btn)
{
    this->verticalLayout->addWidget(btn);
}

void Sidebar::resizeEvent(QResizeEvent *)
{
    // 动画未运行时改变,父窗口高度改变,需重设动画起始和结束高度
    // 这里不判断高度是否相等(性能差不多?? 但是需要增加额外的初始化代码)
    if (!this->isAnimationRunning) //&& this->height()!=this->window()->height()
    {
        this->initialSize.rheight() = this->height();
        this->endSize.rheight() = this->initialSize.rheight();
        this->animation->setStartValue(this->initialSize);
        this->animation->setEndValue(this->endSize);
    }
}

void Sidebar::paintEvent(QPaintEvent *)
{
    QPainter painter(this);
    painter.setBrush(this->backgroundBrush);
    painter.setPen(Qt::PenStyle::NoPen);
    painter.drawRect(this->rect());
    painter.setPen(this->borderLinePen);
    painter.drawLine(this->width(), 0, this->width(), this->height());
}

File: Archive scrap code/PySide6/RetractableSidebarWindow/sidebar.py
from PySide6.QtCore import QPropertyAnimation, QSize, Signal, Qt, Slot, QAbstractAnimation, QRegularExpression
from PySide6.QtGui import qRgb, QBrush, QPen, QPainter
from PySide6.QtWidgets import QWidget, QVBoxLayout


class Sidebar(QWidget):
    expandStart = Signal()  # 开始展开
    expandFinished = Signal()  # 展开结束
    shrinkStart = Signal()  # 开始收缩
    shrinkFinished = Signal()  # 收缩结束
    finished = Signal()  # 展开或收缩结束

    def __init__(self, parent, initial_width: int = 50):
        super().__init__(parent)
        self.isAnimationRunning = False  # 动画是否正在运行
        self.isExpanded = False  # 是否展开
        self.animation = QPropertyAnimation(self)  # 属性动画
        self.animation.setTargetObject(self)
        self.animation.setPropertyName(bytes("size", "utf-8"))
        self.verticalLayout = QVBoxLayout(self)  # 垂直布局
        # self.increasedWidth = 270  # 展开时增加宽度
        self.initialSize = QSize()  # 初始尺寸
        self.endSize = QSize()  # 终止尺寸
        self.backgroundBrush = QBrush(qRgb(243, 243, 243))  # 背景色绘制笔刷
        self.borderLinePen = QPen(qRgb(229, 229, 229))  # 右侧边界线绘制笔

        self.verticalLayout.setSpacing(4)
        self.verticalLayout.setContentsMargins(5, 5, 5, 5)
        self.animation.setDuration(150)
        self.borderLinePen.setWidth(2)

        self.animation.finished.connect(self.animationFinish)

        # 动画起始高度和父窗口高度一样, 宽度为初始宽度, 之后不再更改宽度
        self.initialSize.setWidth(initial_width)
        self.initialSize.setHeight(self.window().height())

        # 动画结束高度和动画起始高度始终相等, 动画结束宽度为动画起始宽度加上展开增加宽度
        self.endSize.setWidth(self.initialSize.width() + 270)
        self.endSize.setHeight(self.initialSize.height())

        self.resize(initial_width, self.window().height())

    def animationFinish(self):
        if self.initialSize.height() != self.window().height():
            # 动画结束, 高度和父窗口不等, 说明动画进行的过程中改变了窗口大小, 需重新设置高度
            self.initialSize.setHeight(self.window().height())
            self.endSize.setHeight(self.initialSize.height())
            self.animation.setStartValue(self.initialSize)
            self.animation.setEndValue(self.endSize)
            self.resize(self.width(), self.window().height())
        self.isExpanded = not self.isExpanded
        self.isAnimationRunning = False
        if self.isExpanded:
            self.expandFinished.emit()
        else:
            self.shrinkFinished.emit()
        self.finished.emit()

    def addWidget(self, widget: QWidget):
        """添加控件"""
        self.verticalLayout.addWidget(widget)

    def addItem(self, item):
        """添加弹簧"""
        self.verticalLayout.addItem(item)

    def setIncreasedWidth(self, increased_width: int):
        """设置展开增加宽度"""
        self.endSize.setWidth(self.endSize.width() + increased_width - self.endSize.width()+self.initialSize.width())

    def setExpandTime(self, ms: int):
        """设置展开时间"""
        self.animation.setDuration(ms)

    def setBackgroundBrush(self, brush: QBrush):
        """设置背景色绘制笔刷"""
        self.backgroundBrush = brush

    def setBorderLinePen(self, pen: QPen):
        """设置右侧边界线绘制笔"""
        self.borderLinePen = pen

    def childrenCumulativeHeight(self) -> int:
        """计算子控件的累计总高度，包括布局边距和控件间距"""
        cumulative_height = 0  # 累计高度
        # 查找直接子控件
        children = self.findChildren(QWidget, QRegularExpression(), Qt.FindChildOption.FindDirectChildrenOnly)
        for child in children:  # 累加直接子控件高度
            cumulative_height += child.height()
        cumulative_height += self.layout().contentsMargins().top()
        cumulative_height += self.layout().contentsMargins().bottom()
        cumulative_height += self.layout().spacing() * (len(children) - 1)
        return cumulative_height

    @Slot()
    def autoExpand(self):
        """自动控制展开或收缩"""
        self.expand()
        self.shrink()

    @Slot()
    def expand(self):
        """展开"""
        if not self.isExpanded and not self.isAnimationRunning:  # 当前不是展开状态
            self.animation.setDirection(QAbstractAnimation.Direction.Forward)  # 正向
            self.animation.start()
            self.isAnimationRunning = True
            self.expandStart.emit()

    @Slot()
    def shrink(self):
        """收缩"""
        if self.isExpanded and not self.isAnimationRunning:
            self.animation.setDirection(QAbstractAnimation.Direction.Backward)  # 逆向
            self.animation.start()
            self.isAnimationRunning = True
            self.shrinkStart.emit()

    def resizeEvent(self, e):
        # 动画未运行时改变, 父窗口高度改变, 需重设动画起始和结束高度
        if not self.isAnimationRunning:
            self.initialSize.setHeight(self.height())
            self.endSize.setHeight(self.initialSize.height())
            self.animation.setStartValue(self.initialSize)
            self.animation.setEndValue(self.endSize)

    def paintEvent(self, e):
        painter = QPainter(self)
        painter.setBrush(self.backgroundBrush)
        painter.setPen(Qt.PenStyle.NoPen)
        painter.drawRect(self.rect())
        painter.setPen(self.borderLinePen)
        painter.drawLine(self.width(), 0, self.width(), self.height())
